### PR TITLE
[5.3] Fixed the validator bug in hydrateFiles when removing the files array

### PR DESCRIPTION
**Pull request by chengguoqiang**
fixed the bug of the method hydrateFiles  when remove the files array.

### CODE REVIEW

**Comment by chengguoqiang** [comment]
tests for #15660

**Comment by themsaid** [comment]
The test you added is currently passing without the changes you introduce, can you please describe what are you trying to fix?

**Comment by chengguoqiang** [comment]
@themsaid  Upload the files array ,  hydrateFiles can't remove it from data array

**Comment by themsaid** [comment]
@chengguoqiang can you please add a failing test?

**Comment by taylorotwell** [comment]
We also need a much better description of what this is attempting to solve / fix / do before merging.

**Comment by chengguoqiang** [comment]
@themsaid I added a failing test, please check, thank you.

**Comment by chengguoqiang** [comment]
@taylorotwell the description was modified, would you please check the description for me

**Comment by themsaid** [comment]
Well you just pointed out a bug, but the fix isn't actually complete. The validator works fine with this bug except for the part of separating data from files for array file uploads, means that now `$this->data` won't have these file uploads, and the `each()` method relys only on the data attribute, so you'll need to update the `each()` method and merge the files array to the `$data` array inside the method before starting the iteration.

To ensure everything is working, you may add these tests:

    public function testFilesHydration()
    {
        $trans = $this->getIlluminateArrayTranslator();
        $file = new File(__FILE__, false);
        $v = new Validator($trans, ['file' => $file, 'text' => 'text'], ['name' => 'Required']);
        $this->assertEquals(['file' => $file], $v->getFiles());
        $this->assertEquals(['text' => 'text'], $v->getData());
    }

    public function testArrayOfFilesHydration()
    {
        $trans = $this->getIlluminateArrayTranslator();
        $file = new File(__FILE__, false);
        $file2 = new File(__FILE__, false);
        $v = new Validator($trans, ['file' => [$file, $file2], 'text' => 'text'], ['name' => 'Required']);
        $this->assertEquals(['file.0' => $file, 'file.1' => $file2], $v->getFiles());
        $this->assertEquals(['text' => 'text'], $v->getData());
    }

    public function testMultipleFileUploads()
    {
        $trans = $this->getIlluminateArrayTranslator();
        $file = new File(__FILE__, false);
        $file2 = new File(__FILE__, false);
        $v = new Validator($trans, ['file' => [$file, $file2]], ['file.*' => 'Required|mimes:xls']);
        $this->assertFalse($v->passes());
    }

    public function testFileUploads()
    {
        $trans = $this->getIlluminateArrayTranslator();
        $file = new File(__FILE__, false);
        $v = new Validator($trans, ['file' => $file], ['file' => 'Required|mimes:xls']);
        $this->assertFalse($v->passes());
    }

**Comment by chengguoqiang** [comment]
@themsaid improvement

**Comment by themsaid** [comment]
I'm sorry but your version of `testMultipleFileUploads()` doesn't really test anything, same for `testFileUploads ()`.

These simple tests should pass:

```
public function testMultipleFileUploads()
{
    $trans = $this->getIlluminateArrayTranslator();
    $file = new File(__FILE__, false);
    $file2 = new File(__FILE__, false);
    $v = new Validator($trans, ['file' => [$file, $file2]], ['file.*' => 'Required|mimes:xls']);
    $this->assertFalse($v->passes());
}

public function testFileUploads()
{
    $trans = $this->getIlluminateArrayTranslator();
    $file = new File(__FILE__, false);
    $v = new Validator($trans, ['file' => $file], ['file' => 'Required|mimes:xls']);
    $this->assertFalse($v->passes());
}
```

**Comment by chengguoqiang** [comment]
I update the each() method and merge the files array to the $data array inside the method before starting the iteration. but these simple tests what you given run fail:

File: src/Illuminate/Validation/Validator.php [hunk 1]
```diff
@@ -2849,9 +2854,9 @@ protected function extractDataFromPath($attribute)
     {
         $results = [];
 
-        $value = Arr::get($this->data, $attribute, '__missing__');
+        $value = $this->getValue($attribute);
 
-        if ($value != '__missing__') {
+        if (! is_null($value)) {
```
Comment (themsaid) [inline src/Illuminate/Validation/Validator.php]: What's the reason for this change? `null` is a value we still want to get.

**Comment by themsaid** [comment]
In the `each()` method you need to have this:

```
$data = array_merge(
            Arr::dot($this->initializeAttributeOnData($attribute)), 
            $this->files
        );
```

so that it looks into the files array as well, if you have this the tests will pass.

**Comment by chengguoqiang** [comment]
the tests pass, thank you very match

**Comment by GrahamCampbell** [comment]
Please rebase and squash to one commtit.
